[llvm] Add triples for managarm

no92 opened the pull request:
This PR aims to add a target for [managarm](https://github.com/managarm/managarm). The targets `{x86_64,aarch64,riscv64}-pc-managarm-{kernel,mlibc}` will be supported.

Discourse RFC: [discourse.llvm.org/t/rfc-new-proposed-managarm-support-for-llvm-and-clang-87845/85884](https://discourse.llvm.org/t/rfc-new-proposed-managarm-support-for-llvm-and-clang-87845/85884)

no92 commented:
Ping

no92 commented:
Ping

no92 commented:
Ping

brad0 commented:
@MaskRay

MaskRay commented on clang/lib/Driver/ToolChains/Gnu.cpp:
```diff
@@ -2562,7 +2566,10 @@ void Generic_GCC::GCCInstallationDetector::AddDefaultGCCPrefixes(
   static const char *const RISCV64Triples[] = {"riscv64-unknown-linux-gnu",
                                                "riscv64-linux-gnu",
-                                               "riscv64-unknown-elf"};
+                                               "riscv64-unknown-elf",
```
L2449 has a comment that we should not add new elements.

The existing ones should be cleaned up as well.

MaskRay commented:
Some older `ToolChain`s were probably contributed with a lot of `CmdArgs.push_back` uncovered by tests. They are not good examples to follow. For new `ToolChain`s, we ensure that all constructed `CmdArgs.push_back` are covered. This allows refactoring by someone who is unfamiliar with your usage pattern.

MaskRay commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -279,8 +280,10 @@ class Triple {
     Amplification,
     OpenCL,
     OpenHOS,
+    Kernel,
```
Why is a generic term `Kernel` added? I am concerned that it would cause confusion to users of other OSes.

Does your OS need a different target triple for kernel development? Note that almost all other OSes don't make the target triple distinction.

MaskRay commented on llvm/include/llvm/TargetParser/Triple.h:
```diff
@@ -279,8 +280,10 @@ class Triple {
     OpenCL,
     OpenHOS,
+    Kernel,
+    Mlibc,
```
I don't know how Mlibc is intended to be used but LLVM LTO warns about differing target triples.

MaskRay commented:
> [llvm] Add triples for managarm

I suggest that you split the patch into LLVM target triple part and a clang part. That's a convention to support new targets.

MaskRay commented:
#78065 for Hurd is a good example for clang testing.

MaskRay commented:
> @MaskRay seems like this target might be too niche to go into LLVM at this time? is it worth considering some bar before accepting such a thing into LLVM, rather than encouraging folks to maintain such a thing in a branch for now?

Good question about the acceptance bar. I started https://discourse.llvm.org/t/new-clang-toolchain-acceptance-criteria/79233

ArsenArsen commented:
(context: I work(ed) on managarm, but haven't interacted with the LLVM port on it nearly at all, and have never added targets to LLVM)

The `Kernel` environment probably was added is to mimic what gnuconfig does:

```
~$ /usr/share/gnuconfig/config.sub x86_64-managarm-kernel
x86_64-pc-managarm-kernel
~$ /usr/share/gnuconfig/config.sub x86_64-linux-kernel
Invalid configuration 'x86_64-linux-kernel': 'linux' does not support 'kernel'.
~ 1 $
```

It might be reasonable to constrain it to `managarm` (as gnuconfig does).

> Does your OS need a different target triple for kernel development? Note that almost all other OSes don't make the target triple distinction.

It does for GCC, at least for libstdc++.  I don't know how LLVM works, so I can't comment there.

The kernel environment could also conceivably be a multilib of `*-*-managarm*-*` in general.  I haven't done the work necessary to implement that in our GCC port yet, but it's been simmering in the back of my mind for a while.  Generally, though, I don't buy that an OS triplet covers said OSes kernel.  It seems simpler for kernel compilers to be kernel compilers (perhaps via multilib).

WRT niche, I don't disagree - that's part of why I haven't upstreamed anything into GCC either.  The other part is that I haven't worked enough on the port to consider myself confident in quality for inclusion (but that doesn't have an influence on the LLVM port)

no92 commented:
Rebased and updated the PR to include the feedback. We've decided to drop the `kernel` environment for now, and will maintain it out-of-tree for the time being. We might come up with a better solution for it in the future, also see @ArsenArsen's comment above for details why we use it currently.

Other than that, the implementation was adapted to work along the same lines as Linux and Hurd - in particular, we don't add triplets to the arrays where we ought not to. Tests were added, and we currently use these patches for working on managarm successfully.

AFAICT the CI failure seems unrelated to the PR, but I have a hard time telling.

no92 commented:
Rebased on main, pinging @MaskRay for re-review.

no92 commented:
Ping

MaskRay commented:
Waiting for a Discourse proposal for Managarm per https://discourse.llvm.org/t/new-clang-toolchain-acceptance-criteria/79233

no92 commented:
Discourse RFC: https://discourse.llvm.org/t/rfc-new-proposed-managarm-support-for-llvm-and-clang-87845/85884

MaskRay reviewed (approved):
The patch mixes the following changes

* Triple
* llvm/lib/Support/Unix
* clang/lib/Driver

I think we need 3 PRs. The Triple and Support code looks good.

I will have limited internet access between April 20th and May 4th, and my response time may be delayed..

MaskRay commented on clang/lib/Basic/Targets/OSTargets.h:
```diff
@@ -395,6 +395,36 @@ class LLVM_LIBRARY_VISIBILITY LinuxTargetInfo : public OSTargetInfo<Target> {
+  void getOSDefines(const LangOptions &Opts, const llvm::Triple &Triple,
+                    MacroBuilder &Builder) const override {
+    DefineStd(Builder, "unix", Opts);
+    Builder.defineMacro("__managarm__");
```
We need a clang/test/Preprocessor test. You might want to do that in a separate PR and add some regular Clang maintainers.

no92 committed:
[llvm] Add managarm OS and mlibc environment target

no92 commented:
Rebased on main. The PR will be split into the three parts as requested, so this one is for the triple only.

brad0 commented:
cc @MaskRay
